### PR TITLE
Default error queue name to "error"

### PR DESCRIPTION
**Pull request by andreasohlund**
But have msmq require an explicitly configured value

Closes https://github.com/Particular/NServiceBus/issues/3647

### CODE REVIEW

**Comment by andreasohlund** [comment]
@timbussmann can you do a review?

File: src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@ public class When_no_custom_configuration_source_is_specified
         {
             var config = new EndpointConfiguration("myendpoint");
 
+            config.SendFailedMessagesTo("error");
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs]: why is this required?

File: src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            {
+                EndpointSetup<PlainVanillaEndpoint>();
+            }
+        }
+
+        public class Context : ScenarioContext
+        {
+        }
+
+        public class PlainVanillaEndpoint : IEndpointSetupTemplate
```
Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs]: that's quite a lot of effort just because the default server configures an error queue :(

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 3]
```diff
@@ -11,6 +11,12 @@ namespace NServiceBus
     public static class ErrorQueueSettings
     {
         /// <summary>
+        /// The settings key where the error queue address is stored.    
+        /// </summary>
+        public const string SettingsKey = "errorQueue";
+        const string DefaultErrorQueueName = "error";
```
Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: could we also set the default value via settings.SetDefault instead?

File: src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Config/When_no_custom_configuration_source_is_specified.cs]: Because we still run the those tests with MSMQ we need that to avoid blowing up. Once the learning transport is merged it can be removed

File: src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs]: I know :( Ideas?

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs (hunk 3 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: We could but that made the code in https://github.com/Particular/NServiceBus/pull/4636/files/25391cbf06b57ba09fe4fe3cdc6c0627e6a74d41#diff-49b73679141b39d739de1e8bdd0d036eR37 harder to grok. Want me to change?

Comment (timbussmann) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: discussed this with @andreasohlund on a call and given his argument about the readability of the code with the current constraints (checking alternative configuration sources for error queue configuration) I'm good with that approach.

File: src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs (hunk 2 above)
Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs]: theoretically we could make a base class `DefaultServer`  inherits from, which does not specify the error queue, but this would be kinda weird too. :(

We could save us the cleanup mechanism as the endpoint should never start in the first place?

**Comment by timbussmann** [comment]
how are we going to document this?

**Comment by andreasohlund** [comment]
>how are we going to document this?

I'll do a WIP doco PR tomorrow

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 4]
```diff
@@ -23,19 +29,31 @@ public static string ErrorQueueAddress(this ReadOnlySettings settings)
...
+            if (TryGetErrorQueueAddress(settings, out errorQueue))
             {
-                Logger.Debug("Error queue retrieved from code configuration via 'EndpointConfiguration.SendFailedMessagesTo().");
                 return errorQueue;
             }
 
+            return DefaultErrorQueueName;
+        }
+
+        internal static bool TryGetErrorQueueAddress(this ReadOnlySettings settings, out string errorQueue)
```
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: This being internal means that MSMQ won't be able to call it when it's moved out of core. Seems like we should go ahead and account for this now. Do we just want to make this public, or do we need to come up with a different approach for MSMQ to get this information?

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 5]
```diff
@@ -23,19 +29,31 @@ public static string ErrorQueueAddress(this ReadOnlySettings settings)
...
+
             var section = settings.GetConfigSection<MessageForwardingInCaseOfFaultConfig>();
             if (section != null)
             {
                 if (!string.IsNullOrWhiteSpace(section.ErrorQueue))
                 {
                     Logger.Debug("Error queue retrieved from <MessageForwardingInCaseOfFaultConfig> element in config file.");
-                    return section.ErrorQueue;
+                    errorQueue = section.ErrorQueue;
+                    return true;
```
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: I can't comment directly on the relevant lines, but should we consider revising the two exception messages in this method? One of the possible actions is to just remove the app.config section/registry key, without needing to add the `SendFailedMessagesTo` call (for everything but MSMQ, of course).

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs (hunk 4 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: Made it public

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs (hunk 5 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: Not sure I follow, you mean to only tell users to use the code first version to prep them for v7 where this will be the only option?

File: src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs]: @timbussmann take a look at https://github.com/Particular/NServiceBus/pull/4636/commits/e70c7270713e6b904f7a0bc20127213da17d971e and see if you like it?

Comment (timbussmann) [inline src/NServiceBus.Msmq.AcceptanceTests/When_no_explicit_error_queue_is_configured.cs]: @andreasohlund 👍 I like it

**Comment by andreasohlund** [comment]
Ok, I've now cleaned up the test

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs (hunk 5 above)
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: As discussed in the meeting, I'm referring to the fact that one of the remediation options is to remove the invalid configuration that caused the exception, and rely on the default error queue value instead.

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 6]
```diff
@@ -23,10 +23,29 @@ public static string ErrorQueueAddress(this ReadOnlySettings settings)
...
 
-            if (settings.TryGet("errorQueue", out errorQueue))
+            if (TryGetExplicitErrorQueueAddress(settings, out errorQueue))
+                return errorQueue;
+
+            return DefaultErrorQueueName;
+        }
+
+        /// <summary>
+        /// Get the explicitly configured error queue address if one is defined.
```
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: Should be "Gets" not "Get"

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 7]
```diff
@@ -23,10 +23,29 @@ public static string ErrorQueueAddress(this ReadOnlySettings settings)
...
+        /// The error queue can be configured in code using 'EndpointConfiguration.SendFailedMessagesTo()',
+        /// via the 'Error' attribute of the 'MessageForwardingInCaseOfFaultConfig' configuration section,
+        /// or using the 'HKEY_LOCAL_MACHINE\SOFTWARE\ParticularSoftware\ServiceBus\ErrorQueue' registry key.
+        /// </summary>
+        /// <param name="settings">The configuration settings of this endpoint.</param>
+        /// <param name="errorQueue">The configured error queue.</param>
+        /// <returns>True if an explict error queue have been configured.</returns>
```
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: Would be better as "True if an error queue has been explicitly configured."

File: src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs [hunk 8]
```diff
@@ -23,10 +23,29 @@ public static string ErrorQueueAddress(this ReadOnlySettings settings)
...
+        /// <exception cref="Exception">When the configuration for the endpoint is invalid.</exception>
+        public static bool TryGetExplicitErrorQueueAddress(this ReadOnlySettings settings, out string errorQueue)
```
Comment (bording) [inline src/NServiceBus.Core/Recoverability/Faults/ErrorQueueSettings.cs]: The naming of this makes it sound like it's an "explicit error queue" vs. an error queue that has been explicitly configured. How about `TryGetExplicitlyConfiguredErrorQueueAddress`? Though that's rather long.

**Comment by andreasohlund** [comment]
@bording your comments should be addressed now, please review

**Comment by bording** [comment]
I made some further revisions to the wording and approved the API. I'll go ahead and merge.
